dbms/src/Flash/Mpp/GRPCSendQueue.h

```diff
@@ -0,0 +1,233 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/Exception.h>
+#include <Common/Logger.h>
+#include <Common/MPMCQueue.h>
+#include <common/logger_useful.h>
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#endif
+#include <grpcpp/grpcpp.h>
+#ifdef __clang__
+#pragma clang diagnostic pop
+#endif
+
+#include <functional>
+
+namespace DB
+{
+
+namespace tests
+{
+class TestGRPCSendQueue;
+} // namespace tests
+
+using GRPCKickFunc = std::function<grpc_call_error(void *)>;
+
+enum class GRPCSendQueueRes
+{
+    OK,
+    FINISHED,
+    EMPTY,
+};
+
+/// A multi-producer-single-consumer queue dedicated to async grpc streaming send work.
+///
+/// In streaming rpc, a client/server may send messages continuous.
+/// However, async grpc is only allowed to have one outstanding write on the
+/// same side of the same stream without waiting for the completion queue.
+/// Further more, the message usually is generated from another thread which
+/// introduce a race between this thread and grpc threads.
+/// The grpc cpp framework provides a tool named `Alarm` can be used to push a tag into
+/// completion queue thus the write can be done in grpc threads. But `Alarm` must need
+/// a timeout and it uses a timer to trigger the notification, which is wasteful if we want
+/// to trigger it immediately. So we can say `kickCompletionQueue` function is a
+/// immediately-triggered `Alarm`.
+template <typename T>
+class GRPCSendQueue
+{
+public:
+    GRPCSendQueue(size_t queue_size, grpc_call * call, const LoggerPtr log_)
+        : send_queue(MPMCQueue<T>(queue_size))
+        , tag(nullptr)
+        , log(log_)
+    {
+        RUNTIME_ASSERT(call != nullptr, log, "call is null");
+        // If a call to `grpc_call_start_batch` with an empty batch returns
+        // `GRPC_CALL_OK`, the tag is pushed into the completion queue immediately.
+        // This behavior is well-defined. See https://github.com/grpc/grpc/issues/16357.
+        kick_func = [call](void * t) {
+            return grpc_call_start_batch(call, nullptr, 0, new KickTag(t), nullptr);
```

Comment:
I had some concerns about here. Then I realize that when `tag` is not `nullptr` the EstablishCallData is stopped and only waiting on notification form here. So there won't be any race.

How about mentions this in comment?

Reply:
I add more comments above `pop` because the `tag` is passed through it.